crates/warg-server/src/services/core.rs

```diff
@@ -0,0 +1,414 @@
+use anyhow::Error;
+use serde::{Deserialize, Serialize};
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+use tokio::sync::mpsc::{self, Receiver, Sender};
+use tokio::sync::oneshot;
+use tokio::task::JoinHandle;
+use warg_crypto::hash::{Digest, DynHash, Hash, Sha256};
+use warg_protocol::registry::LogLeaf;
+use warg_protocol::{
+    operator, package,
+    registry::{LogId, MapCheckpoint, RecordId},
+    Envelope,
+};
+
+#[derive(Clone, Debug, Default)]
+pub struct State {
+    checkpoints: Vec<Arc<Envelope<MapCheckpoint>>>,
+    operator_state: Arc<Mutex<OperatorInfo>>,
+    package_states: HashMap<LogId, Arc<Mutex<PackageInfo>>>,
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+struct OperatorInfo {
+    validator: operator::Validator,
+    log: Vec<Arc<Envelope<operator::OperatorRecord>>>,
+    records: HashMap<RecordId, OperatorRecordInfo>,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct OperatorRecordInfo {
+    record: Arc<Envelope<operator::OperatorRecord>>,
+    state: RecordState,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct PackageInfo {
+    id: LogId,
+    name: String,
+    validator: package::Validator,
+    log: Vec<Arc<Envelope<package::PackageRecord>>>,
+    records: HashMap<RecordId, PackageRecordInfo>,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct PackageRecordInfo {
+    pub record: Arc<Envelope<package::PackageRecord>>,
+    pub content_sources: Arc<Vec<ContentSource>>,
+    pub state: RecordState,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct ContentSource {
+    pub content_digest: DynHash,
+    pub kind: ContentSourceKind,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(tag = "type")]
+pub enum ContentSourceKind {
+    HttpAnonymous { url: String },
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
+pub enum RecordState {
+    #[default]
+    Unknown,
+    Processing,
+    Published {
+        checkpoint: Arc<Envelope<MapCheckpoint>>,
+    },
+    Rejected {
+        reason: String,
+    },
+}
+
+pub struct CoreService {
+    mailbox: mpsc::Sender<Message>,
+    handle: JoinHandle<State>,
```

Comment:
If the goal is to persist to a db anyway, maybe strip some of this graceful shutdown complexity?